### PR TITLE
Revert "Separate TransformerEmbedding layer (#33)"

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #214

Avoid diverging the model structure (FQNs and checkpoint
interoperability) with similar models.

This reverts commit b99af33543497a4d592b20c7033023d18a4ea660.

### CODE REVIEW

File: torchtrain/models/llama/model.py [hunk 1]
```diff
@@ -422,7 +396,17 @@ def init_weights(self):
...
-        self.embeddings.init_weights()
+
+        self.freqs_cis = precompute_freqs_cis(
+            self.model_args.dim // self.model_args.n_heads,
+            # Need to compute until at least the max token limit for generation
+            # (use 2x max sequence length to be safe)
+            self.model_args.max_seq_len * 2,
+        )
+
+        with torch.device(self.freqs_cis.device):
```
Comment (tianyu-l) [inline torchtrain/models/llama/model.py:395]: IIUC, `with torch.device(self.freqs_cis.device):` should wrap `self.freqs_cis =` init, not `self.tok_embeddings` init.
The reason is that after meta init and to(cuda), `tok_embeddings` is already on cuda, and we use this context to make sure `precompute_freqs_cis` is generating new buffer on cuda.

Comment (wconstab) [reply torchtrain/models/llama/model.py:395]: ok fixed that.

File: torchtrain/models/llama/model.py [hunk 2]
```diff
@@ -394,11 +346,33 @@ class Transformer(nn.Module):
...
+        # TODO persistent should be set to false, since this buffer can be recomputed.
+        # however, we set it to true for 2 reasons.  (1) due to pytorch/pytorch#123411,
+        # compile or pipeline-tracer will not correctly handle non-persistent buffers,
+        # so we need to fix that.  (2) if we initialize pipeline-parallel models from
+        # a seed checkpoint rather than calling init_weights, we need freqs_cis to be
+        # initialized by the checkpoint, or we need to add a separate initializer for
+        # just the non-persistent buffers that is called after loading checkpoints.
+
+        # Note that self.params.max_seq_len is multiplied by 2 because the token limit for the Llama 2 generation of models is 4096.
+        # Adding this multiplier instead of using 4096 directly allows for dynamism of token lengths while training or fine-tuning.
```
Comment (awgu) [inline torchtrain/models/llama/model.py]: Let us delete this comment. It is copied from the original Llama release, but I think even there, the comment is not accurate. The comment below it should be accurate.

File: torchtrain/models/llama/model.py [hunk 3]
```diff
@@ -422,7 +396,17 @@ def init_weights(self):
         ``init_weights``. We only call it in the constructor of this
         ``Transformer`` root module to avoid reinitializing tensors.
         """
-        self.embeddings.init_weights()
+
```
Comment (awgu) [inline torchtrain/models/llama/model.py]: nit: no need for these extra newlines here and around the added code (for consistency with the rest of the `init_weights` method)

File: torchtrain/models/llama/model.py [hunk 4]
```diff
@@ -394,11 +346,33 @@ class Transformer(nn.Module):
     def __init__(self, model_args: ModelArgs):
         super().__init__()
         self.model_args = model_args
+        self.model_dim = model_args.dim
```
Comment (awgu) [inline torchtrain/models/llama/model.py]: If we save this attribute, shall we use it in place of `self.model_args.dim` below? Or, should we just not save this attribute at all?

Comment (wconstab) [reply torchtrain/models/llama/model.py]: ok- i went with not saving this attribute, although, oddly it was being used below in one place which i updated.

File: torchtrain/models/llama/model.py [hunk 5]
```diff
@@ -396,9 +348,27 @@ def __init__(self, model_args: ModelArgs):
         self.model_args = model_args
         self.vocab_size = model_args.vocab_size
         self.n_layers = model_args.n_layers
-        self.model_dim = model_args.dim
 
-        self.embeddings = RotaryEmbedding(model_args)
+        self.tok_embeddings = nn.Embedding(model_args.vocab_size, model_args.dim)
```
Comment (wanchaol) [inline torchtrain/models/llama/model.py:352]: I think we will also need to fix the TP plan too in this PR, it was wrapped in a wrapper module but now we need to remove that wrapper fqn

Comment (wanchaol) [reply torchtrain/models/llama/model.py:352]: specifically here https://github.com/pytorch/torchtitan/blob/main/torchtrain/parallelisms/parallelize_llama.py#L157

Comment (wconstab) [reply torchtrain/models/llama/model.py:352]: fixed
